### PR TITLE
Support Cross-Platform CPU and Memory counters.

### LINKED ISSUE #1189: Support Process CPU and Memory from Linux
Overall story:
https://github.com/microsoft/ApplicationInsights-dotnet-server/issues/1188

The goal is to leverage the following to report process cpu, process memory usage. 
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.totalprocessortime
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.privatememorysize64

These 2 counters are what powers LiveMetric charts.(which appear blank in Linux now)
image

### PR DESCRIPTION
**Pull request by cijothomas**
Fix #1189 

@SergeyKanzhelev @Dmitry-Matveev @lmolkova @MS-TimothyMothra Please share your thoughts.

Some highlights:
1. XPlatform counters currently support just 3 - process cpu, process cpu normalized, process/private bytes.
2. There are no public API changes - users still configure counters as before. Underneath we use StandardPerfCollection or Web App or something else like In-Proc stats, EventCounters - users are abstracted from this as of this PR. (this will change soon when we start supporting more counters.)
3. I have manually tested this in windows and linux. LiveMetrics also work in Linux now, showing process cpu/memory.

 Will eventually enable Linux CI builds for this repo to get automatic tests.

### CODE REVIEW

File: Src/PerformanceCollector/Perf.Shared.NetStandard20Net45/StandardPerformanceCollector/StandardPerformanceCollector.cs [hunk 1]
```diff
@@ -94,6 +94,7 @@ public void RefreshCounters()
                 perfCounterName,
                 this.win32Instances,
                 this.clrInstances,
+                true,
```
Comment (MS-TimothyMothra) [inline Src/PerformanceCollector/Perf.Shared.NetStandard20Net45/StandardPerformanceCollector/StandardPerformanceCollector.cs]: nit: consider using parameter names here. helps the reader understand what the "true" value is set to.

File: Src/PerformanceCollector/Perf.Shared/Implementation/PerformanceCounterUtility.cs [hunk 2]
```diff
@@ -58,6 +68,100 @@ public static string FormatPerformanceCounter(PerformanceCounter pc)
...
+                else
+                {
+                    // return false;
+                    return true;
+                }
+#endif
+            }
+        }
+
+#if NETSTANDARD1_6
```
Comment (MS-TimothyMothra) [inline Src/PerformanceCollector/Perf.Shared/Implementation/PerformanceCounterUtility.cs]: nit: I see that for the method `GetPerformanceCollector`, NETSTANDARD1.6 has it's own method, but NET45 and NETSTANDARD2.0 are sharing a method.

I would expect that either all three have separate methods or all three share the same method.

**Comment by MS-TimothyMothra** [comment]
I see a lot of methods with platform preprocessors (NET45, NETSTANDARD, etc).
I wonder if we would be better off writing a factory that returns a specific class based on framework.

what do you think?

File: Src/PerformanceCollector/Perf.Shared/Implementation/PerformanceCounterUtility.cs (hunk 2 above)
Comment (cijothomas) [reply Src/PerformanceCollector/Perf.Shared/Implementation/PerformanceCounterUtility.cs]: right. moving all to separate methods. its much easier to read now.

File: Src/PerformanceCollector/Perf.Shared.NetStandard20/XPlatProcessCPUPerformanceCounterNormalized.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    {        
+        private readonly bool isInitialized = false;
+        private readonly int processorsCount = -1;
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="XPlatProcessCPUPerformanceCounterNormalized" /> class.
+        /// </summary>
+        internal XPlatProcessCPUPerformanceCounterNormalized() : base()
+        {            
+            this.processorsCount = Environment.ProcessorCount;
```
Comment (Dmitry-Matveev) [inline Src/PerformanceCollector/Perf.Shared.NetStandard20/XPlatProcessCPUPerformanceCounterNormalized.cs]: There is a code snippet below where Environment.ProcessorCount is in Try/Catch - do we need the same here?

File: Src/PerformanceCollector/Perf.Shared.NetStandard20/XPlatProcessCPUPerformanceCounter.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                double previouslyCollectedValue = this.lastCollectedValue;
+                this.lastCollectedValue = Process.GetCurrentProcess().TotalProcessorTime.Ticks;
+
+                var previouslyCollectedTime = this.lastCollectedTime;
+                this.lastCollectedTime = DateTimeOffset.UtcNow;
+
+                double value = 0;
+                if (previouslyCollectedTime != DateTimeOffset.MinValue)
+                {
+                    var baseValue = this.lastCollectedTime.Ticks - previouslyCollectedTime.Ticks;
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply Src/PerformanceCollector/Perf.Shared.NetStandard20/XPlatProcessCPUPerformanceCounter.cs]: lets be safe. will handle it to report error, and report 0.

File: Src/PerformanceCollector/Perf.NetFull.Tests/Perf.NetFull.Tests.csproj [hunk 5]
```diff
@@ -24,14 +24,14 @@
     <DebugSymbols>true</DebugSymbols>
     <DebugType>full</DebugType>
     <Optimize>false</Optimize>
-    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <DefineConstants>DEBUG;TRACE;NET45</DefineConstants>
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply Src/PerformanceCollector/Perf.NetFull.Tests/Perf.NetFull.Tests.csproj]: no diff between 45 and 46

File: Src/PerformanceCollector/Perf.Shared.NetStandard20/XPlatProcessCPUPerformanceCounterNormalized.cs (hunk 3 above)
Comment (cijothomas) [reply Src/PerformanceCollector/Perf.Shared.NetStandard20/XPlatProcessCPUPerformanceCounterNormalized.cs]: good catch. moved to use the utility method.

File: Src/PerformanceCollector/Perf.Shared.NetStandard20/XPlatProcessMemoryPerformanceCounter.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    {        
+        /// <summary>
+        /// Returns the current value of the counter as a <c ref="MetricTelemetry"/>.
+        /// </summary>
+        /// <returns>Value of the counter.</returns>
+        public double Collect()
+        {
+            try
+            {
+                return Process.GetCurrentProcess().PrivateMemorySize64;
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply Src/PerformanceCollector/Perf.Shared.NetStandard20/XPlatProcessMemoryPerformanceCounter.cs]: From doc: This property can be used to monitor memory usage on computers with 32-bit processors or 64-bit processors. The property value is equivalent to the Private Bytes performance counter for the process.

We catch exception to handle unsupported platforms anyway.

File: Src/PerformanceCollector/Perf.Shared/Implementation/PerformanceCollectorEventSource.cs [hunk 7]
```diff
@@ -180,8 +180,24 @@ public void PerfCounterNetCoreOnlyOnAzureWebApp(string applicationName = "dummy"
             this.WriteEvent(21, this.applicationNameProvider.Name);
         }
 
+        [Event(23, Keywords = Keywords.UserActionable, Level = EventLevel.Error, Message = @"Performance counter is not available in the supported list of XPlatform counters. Counter is {0}.")]
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply Src/PerformanceCollector/Perf.Shared/Implementation/PerformanceCollectorEventSource.cs]: fixed.

File: Src/PerformanceCollector/Perf.Shared/Implementation/PerformanceCounterUtility.cs [hunk 8]
```diff
@@ -58,6 +70,91 @@ public static string FormatPerformanceCounter(PerformanceCounter pc)
...
+            {
+                // The original Windows PerformanceCounter collector which is also
+                // supported in NetStandard2.0 in Windows.
+                collector = (IPerformanceCollector)new StandardPerformanceCollector();
+                PerformanceCollectorEventSource.Log.InitializedWithCollector(collector.GetType().Name);
+            }
+            else
+            {
+                // This is NetStandard2.0 and non-windows. Use XPlatformPerfCollector
+                collector = (IPerformanceCollector)new PerformanceCollectorXPlatform();
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply Src/PerformanceCollector/Perf.Shared/Implementation/PerformanceCounterUtility.cs]: I will do more refactoring in the Gauge area to re-use this. In general, the answer to your question is Yes - we should simply use the XPlat mechanism in all the places. Will address in followup PR. Opening issue to track.

**Pull request merged**